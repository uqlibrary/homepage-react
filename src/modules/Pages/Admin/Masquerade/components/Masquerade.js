import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid';
import { styled } from '@mui/material/styles';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';

import { StandardCard } from 'modules/SharedComponents/Toolbox/StandardCard';
import { StandardPage } from 'modules/SharedComponents/Toolbox/StandardPage';
import txt from './masquerade.locale';

const StyledStandardCard = styled(StandardCard)(() => ({
    backgroundColor: 'white',
}));

const Masquerade = ({ account }) => {
    const [userName, setUserName] = useState('');
    const [loading, setLoading] = useState(false);

    const masqueradeAs = event => {
        if ((event && event.key && event.key !== 'Enter') || userName.length === 0) return;

        setLoading(true);

        sessionStorage.removeItem('userAccount');

        let redirectUrl = `${window.location.protocol}//${window.location.hostname}`;
        const isDevBranch = window.location.hostname === 'homepage-development.library.uq.edu.au';
        redirectUrl += isDevBranch ? /* istanbul ignore next */ window.location.pathname : '/';
        window.location.assign(
            `https://auth.library.uq.edu.au/masquerade?user=${userName}&return=${window.btoa(redirectUrl)}`,
        );
    };

    const usernameChanged = event => {
        setUserName(event.target.value);
    };

    const userCapabilityStatement = canMasqueradeType => {
        return canMasqueradeType && canMasqueradeType === 'readonly'
            ? txt.user.access.readonly.capabilityStatement
            : txt.user.access.full.capabilityStatement;
    };

    return (
        <StandardPage title={txt.title}>
            <StyledStandardCard noHeader standardCardId="masquerade">
                <Typography>{userCapabilityStatement(account.canMasqueradeType)}</Typography>
                <Grid container spacing={3} alignItems={'flex-end'} style={{ marginTop: 12 }}>
                    <Grid item xs>
                        <TextField
                            variant="standard"
                            data-testid="masquerade-userName"
                            fullWidth
                            label={txt.labels.hint}
                            value={userName}
                            onChange={usernameChanged}
                            InputProps={{ onKeyPress: masqueradeAs }}
                        />
                    </Grid>
                    <Grid item xs={12} sm={'auto'}>
                        <Button
                            variant="contained"
                            id="submitMasquerade"
                            data-analyticsid="submitMasquerade"
                            data-testid="masquerade-submit"
                            fullWidth
                            color="primary"
                            children={txt.labels.submit}
                            disabled={loading}
                            onClick={masqueradeAs}
                            onKeyUp={masqueradeAs}
                        />
                    </Grid>
                </Grid>
            </StyledStandardCard>
        </StandardPage>
    );
};

Masquerade.propTypes = {
    account: PropTypes.object.isRequired,
};

export default React.memo(Masquerade);
