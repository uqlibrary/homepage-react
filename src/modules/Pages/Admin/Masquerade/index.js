export { default as Masquerade } from './containers/Masquerade';
