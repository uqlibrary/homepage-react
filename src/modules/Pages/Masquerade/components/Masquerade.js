import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import txt from './masqueradeLocale';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import { StandardCard } from 'modules/SharedComponents/Toolbox/StandardCard';
import { StandardPage } from 'modules/SharedComponents/Toolbox/StandardPage';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';

export default class Masquerade extends PureComponent {
    static propTypes = {
        account: PropTypes.object.isRequired,
    };

    constructor(props) {
        super(props);
        this.state = {
            userName: '',
            loading: false,
        };
    }

    _masqueradeAs = event => {
        if ((event && event.key && event.key !== 'Enter') || this.state.userName.length === 0) return;

        this.setState({
            loading: true,
        });

        const redirectUrl = `${window.location.protocol}//${window.location.hostname}/`;
        window.location.assign(
            `https://auth.library.uq.edu.au/masquerade?user=${this.state.userName}&return=${window.btoa(redirectUrl)}`,
        );
    };

    _usernameChanged = event => {
        this.setState({
            userName: event.target.value,
        });
    };

    userCapabilityStatement = canMasqueradeType => {
        return canMasqueradeType && canMasqueradeType === 'readonly'
            ? txt.user.access.readonly.capabilityStatement
            : txt.user.access.full.capabilityStatement;
    };

    render() {
        return (
            <StandardPage title={txt.title} goBackFunc={() => history.back()} goBackTooltip="Go back">
                <div className="layout-card">
                    <StandardCard noHeader>
                        <Grid container spacing={1}>
                            <Grid item xs={12} id="masquerade">
                                <Typography>
                                    {this.userCapabilityStatement(this.props.account.canMasqueradeType)}
                                </Typography>
                                <Grid container spacing={3} alignItems={'flex-end'} style={{ marginTop: 12 }}>
                                    <Grid item xs>
                                        <TextField
                                            data-testid="masquerade-userName"
                                            fullWidth
                                            id="userName"
                                            label={txt.labels.hint}
                                            onChange={this._usernameChanged}
                                            onKeyPress={this._masqueradeAs}
                                            value={this.state.userName}
                                        />
                                    </Grid>
                                    <Grid item xs={12} sm={'auto'}>
                                        <Button
                                            children={txt.labels.submit}
                                            color="primary"
                                            data-testid="masquerade-submit"
                                            disabled={this.state.loading}
                                            fullWidth
                                            id="submitMasquerade"
                                            onClick={this._masqueradeAs}
                                            variant="contained"
                                        />
                                    </Grid>
                                </Grid>
                            </Grid>
                        </Grid>
                    </StandardCard>
                </div>
            </StandardPage>
        );
    }
}
