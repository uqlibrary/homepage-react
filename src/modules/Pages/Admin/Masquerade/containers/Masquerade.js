import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router-dom';
import * as actions from 'actions';

import Masquerade from '../components/Masquerade';

const mapStateToProps = state => {
    return {
        ...state.get('accountReducer'),
    };
};

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(actions, dispatch),
    };
}

let MasqueradeContainer = connect(mapStateToProps, mapDispatchToProps)(Masquerade);
MasqueradeContainer = withRouter(MasqueradeContainer);

export default MasqueradeContainer;
