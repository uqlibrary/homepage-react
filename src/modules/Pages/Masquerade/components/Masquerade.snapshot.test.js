import Masquerade from './Masquerade';

function setup(testProps = {}) {
    const props = {
        author: testProps.author || null,
        actions: testProps.actions || {},
        history: testProps.history || {
            push: jest.fn(),
        },
        account: testProps.account || {},
        ...testProps,
    };
    return getElement(Masquerade, props);
}

describe('Component Masquerade', () => {
    it('Should render form as expected', () => {
        const wrapper = setup();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should render readonly description', () => {
        const props = {
            account: { canMasqueradeType: 'readonly' },
        };
        const wrapper = setup(props);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
