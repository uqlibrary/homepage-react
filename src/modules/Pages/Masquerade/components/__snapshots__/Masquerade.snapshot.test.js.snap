// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Masquerade Should render form as expected 1`] = `
<StandardPage
  goBackFunc={[Function]}
  goBackTooltip="Go back"
  title="Masquerade"
>
  <div
    className="layout-card"
  >
    <StandardCard
      noHeader={true}
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        spacing={1}
      >
        <WithStyles(ForwardRef(Grid))
          id="masquerade"
          item={true}
          xs={12}
        >
          <WithStyles(ForwardRef(Typography))>
            <span>
              <strong>
                WARNING!!
              </strong>
               When masquerading as a user, you will effectively become that user, and changes you make will apply to the account!
            </span>
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Grid))
            alignItems="flex-end"
            container={true}
            spacing={3}
            style={
              Object {
                "marginTop": 12,
              }
            }
          >
            <WithStyles(ForwardRef(Grid))
              item={true}
              xs={true}
            >
              <WithStyles(ForwardRef(TextField))
                data-testid="masquerade-userName"
                fullWidth={true}
                id="userName"
                label="Enter a UQ staff or student username (eg. uqjsmith1 or s123456)"
                onChange={[Function]}
                onKeyPress={[Function]}
                value=""
              />
            </WithStyles(ForwardRef(Grid))>
            <WithStyles(ForwardRef(Grid))
              item={true}
              sm="auto"
              xs={12}
            >
              <WithStyles(ForwardRef(Button))
                color="primary"
                data-testid="masquerade-submit"
                disabled={false}
                fullWidth={true}
                id="submitMasquerade"
                onClick={[Function]}
                variant="contained"
              >
                Masquerade
              </WithStyles(ForwardRef(Button))>
            </WithStyles(ForwardRef(Grid))>
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </StandardCard>
  </div>
</StandardPage>
`;

exports[`Component Masquerade Should render readonly description 1`] = `
<StandardPage
  goBackFunc={[Function]}
  goBackTooltip="Go back"
  title="Masquerade"
>
  <div
    className="layout-card"
  >
    <StandardCard
      noHeader={true}
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        spacing={1}
      >
        <WithStyles(ForwardRef(Grid))
          id="masquerade"
          item={true}
          xs={12}
        >
          <WithStyles(ForwardRef(Typography))>
            <span>
              <strong>
                NOTE:
              </strong>
               As a read-only masquerader, you can view all parts of the profile, but you are not able to make any changes to the account.
            </span>
          </WithStyles(ForwardRef(Typography))>
          <WithStyles(ForwardRef(Grid))
            alignItems="flex-end"
            container={true}
            spacing={3}
            style={
              Object {
                "marginTop": 12,
              }
            }
          >
            <WithStyles(ForwardRef(Grid))
              item={true}
              xs={true}
            >
              <WithStyles(ForwardRef(TextField))
                data-testid="masquerade-userName"
                fullWidth={true}
                id="userName"
                label="Enter a UQ staff or student username (eg. uqjsmith1 or s123456)"
                onChange={[Function]}
                onKeyPress={[Function]}
                value=""
              />
            </WithStyles(ForwardRef(Grid))>
            <WithStyles(ForwardRef(Grid))
              item={true}
              sm="auto"
              xs={12}
            >
              <WithStyles(ForwardRef(Button))
                color="primary"
                data-testid="masquerade-submit"
                disabled={false}
                fullWidth={true}
                id="submitMasquerade"
                onClick={[Function]}
                variant="contained"
              >
                Masquerade
              </WithStyles(ForwardRef(Button))>
            </WithStyles(ForwardRef(Grid))>
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </StandardCard>
  </div>
</StandardPage>
`;
